doc/configuration.rst
Configuration
=============


In your project :

* settings

::

    INSTALLED_APPS = [
        ...
        # apps required by CRUD
        'geostore',  # store geographic data
        'template_model',  # store template in model
        'template_engines',  # generate odt and docx templates
        'rest_framework',  # if you want to try api HTML interface
        'django_json_widget',  # if you want to use django admin
        'reversion',  # used to store every change on data (run ./manage.py createinitialrevisions first)
        # CRUD app
        'terra_geocrud',
        ...
    ]
    ...
    TEMPLATES = [
        ...
        # if you want to render odt templates
        {'BACKEND': 'template_engines.backends.odt.OdtEngine'},
        # if you want to render docx templates
        {'BACKEND': 'template_engines.backends.docx.DocxEngine'},
    ]

* urls

::

    urlpatterns = [
        ...
        path('api/mapbox-baselayer/', include('mapbox_baselayer.urls')),
        path('api/crud/', include('terra_geocrud.urls')),
        ...
    ]

You can customize default url and namespace by including terra_geocrud.views directly

Run migrations

::

    ./manage.py migrate



- ADMIN :

you can disable and / or customize admin

::
from django.contrib import admin
from geostore.models import Layer, Feature
from mapbox_baselayer.admin import MapBaseLayerAdmin
from mapbox_baselayer.models import MapBaseLayer

from terra_geocrud import admin as geocrud_admin, models

admin.site.register(models.CrudGroupView, geocrud_admin.CrudGroupViewAdmin)
admin.site.register(models.CrudView, geocrud_admin.CrudViewAdmin)
admin.site.register(models.AttachmentCategory, geocrud_admin.AttachmentCategoryAdmin)

# we recommend to activate MapBaseLayerAdmin to manage map base layers
admin.site.register(MapBaseLayer, MapBaseLayerAdmin)

# we recommend to use thes admin for this geostore models
admin.site.register(Layer, geocrud_admin.CrudLayerAdmin)
admin.site.register(Feature, geocrud_admin.CrudFeatureAdmin)



- SETTINGS :

Waiting for settings definition directly in models.

Settings should be overrided  with TERRA_GEOCRUD settings in your project settings file:

::

    ...
    TERRA_GEOCRUD = {
        # default value for map extent. API serialize this for layer extent if there is no features in it (as default)
        'EXTENT': [-90.0, -180.0, 90.0, 180.0],
        # default storage for file stored in json properties. It is recommended to configure a private web storage in your project (as S3Storage -> see django-storages)
        'DATA_FILE_STORAGE_CLASS': 'django.core.files.storage.FileSystemStorage',
        # default mapbox style provided by api if no custom style defined in crud view
        'STYLES': {
            'line': {
                'type': 'line',
                'paint': {
                    'line-color': '#000',
                    'line-width': 3
                }
            },
            'point': {
                'type': 'circle',
                'paint': {
                    'circle-color': '#000',
                    'circle-radius': 8
                }
            },
            'polygon': {
                'type': 'fill',
                'paint': {
                    'fill-color': '#000'
                }
            },
        }
    }
    ...

* If you want to generate map on your template with the geometry of your feature, and/or extra features, you should use
  mbglrenderer.

  Check https://github.com/consbio/mbgl-renderer.

  Change the url in the settings to use your instance of mbglrenderer :

::

    MBGLRENDERER_URL = 'http://mbglrenderer'


CHANGES.md

CHANGELOG
=========

0.3.44.dev0     (XXXX-XX-XX)
----------------------------

* Terra Accounts User Model is not required anymore
* Dont clean features values when property is deleted. (Need to clean with admin crud view action)

0.3.43          (2020-05-27)
----------------------------

* Fix case with date format

0.3.42          (2020-05-27)
----------------------------

* Format date in display values
* Fix default cases when layer has no schema

0.3.41          (2020-05-26)
----------------------------

* Fix mandatory field in crud view admin
* Fix feature cleaning method

0.3.40          (2020-05-25)
----------------------------

* Gdal supported file format to import data in admin

0.3.39          (2020-05-15)
----------------------------

* Update translations
* Fix deprecation and resource warnings
* Delete thumbnail to image deletion
* Delete media files to image deletion


0.3.38.1        (2020-04-24)
----------------------------

* fix translations
* fix feature properties cleaning


0.3.38          (2020-04-23)
----------------------------

* validate json schema properties
* Disable property value affectation at creation


0.3.37          (2020-04-20)
----------------------------

* improve thumbnail generation
* change api default order for attachments and pictures


0.3.36          (2020-04-17)
----------------------------

* Helper to sort elements in django admin


0.3.35          (2020-04-17)
----------------------------

* Fix way to generate templates
* Add local server time to generated files from templates
* Improve settings with supported image formats and max upload file size in bytes
* Now all properties are managed directly in Crud View admin
* Feature detail api endpoint improved


0.3.34          (2020-03-20)
----------------------------
## Bug fixes

* Fix image url in array
* Try to fix some thumbnail generation

## Features

* Add all geometries description in feature detail serializer
* Change generated document name with feature title
* Custom serializer for feature extra geometries
* Fix old serializer feature properties
* Manage plural names for crud views

## Deprecate

* feature serializer new_display_properties became display_properties
* extra_geometries is no longer available
* Widget for data rendering is no longer available


0.3.33          (2020-02-14)
----------------------------

* BREAKING CHANGES : new way to store path in storage
* Fix image generation from data stored image
* New template tag to handle image from data-url stored image
* Fix bug in admin

0.3.32          (2020-02-06)
----------------------------

* Back to django 3.0 compatibility
* New serializer detail to provide features properties informations and data


0.3.31          (2020-01-29)
----------------------------

* Fix compatibility with geostore 0.3.16


0.3.30          (2020-01-27)
----------------------------

* back from django 2.2 maxi. Wait for sorl-thumbnail 12.6.0


0.3.29          (2020-01-27)
----------------------------

* Support Django 3.0
* Compatibility with geostore 0.3.16


0.3.28          (2019-12-17)
----------------------------

* Django Rest Framework 3.11 compatibility
* Python 3.8 compatibility


0.3.27          (2019-12-11)
----------------------------

* Fake data-url content to decrease feature json property size


0.3.26          (2019-12-05)
----------------------------

* add extra geometries identifiers in feature detail endpoint


0.3.25          (2019-12-04)
----------------------------

* HotFix custom styles management


0.3.24          (2019-12-04)
----------------------------

* add crud view property to describe available layers for layer / feature (extra) geometries


0.3.23          (2019-12-03)
----------------------------

* fix settings with only MAX_ZOOM for map style
* Add map_image_url_loader tag  allowing to add map with style, extra_features
* add property to generated mapbox style


0.3.22          (2019-11-13)
----------------------------

* ability to hide ungrouped views in config menu

0.3.21          (2019-11-04)
----------------------------

* Add default widget to render array object as table


0.3.20          (2019-10-21)
----------------------------

* Picture and attachments are not grouped anymore


0.3.19          (2019-10-21)
----------------------------

* Picture and attachments are now behind feature
* Split public / private settings


0.3.18          (2019-10-18)
----------------------------

* Improve integration in terralego settings system
* Manage attachments and pictures to features


0.3.17          (2019-10-17)
----------------------------

News:

* Admin is not longer activated by default. Please configure in your project


0.3.16          (2019-10-15)
----------------------------

Fixes:

* fix admin with custom geostore admin

News:

* Add french translations

0.3.15          (2019-10-14)
----------------------------

* fix missing permissions


0.3.14          (2019-10-14)
----------------------------

* Fix default extent as simple array
* Use django-reversion to keep changes in admin
* Add functionnal permissions (used only in frontend for the moment)


0.3.13          (2019-10-11)
----------------------------

Fixes

* map_style is now empty and not null in case of undefined geometry layer
* extent for layer with no features


0.3.12          (2019-10-11)
----------------------------

Update

* Optimize widget rendering


0.3.11          (2019-10-10)
----------------------------

Update

* Optimize data file storage


0.3.10          (2019-10-09)
----------------------------

News

* Store and serve data file through a customizable django storage (FileSystem / Public by default)

Fixes

* Fix dict iteration in feature creation


0.3.9           (2019-10-09)
----------------------------

Fixes

* Fix feature creation with grouped properties

News

* Add Date format property render widget.


0.3.8           (2019-10-08)
----------------------------

News

* Add json editor in django admin

terra_geocrud/settings.py
from copy import deepcopy

from django.conf import settings

_DEFAULT_TERRA_GEOCRUD = {
    # default extent to world
    'EXTENT': [-90.0, -180.0, 90.0, 180.0],
    'DATA_FILE_STORAGE_CLASS': 'django.core.files.storage.FileSystemStorage',
    # all PIL formats are allowed, but to be used in odt template, should match
    # https://wiki.openoffice.org/wiki/FR/Documentation/Writer_Guide/Types_fichier_image
    'IMAGE_EXTENSION_ALLOWED': ['.jpeg', '.jpg', '.png', '.gif', '.bmp'],
    # give max_memory_size allowed by backend to api settings
    'FILE_UPLOAD_MAX_MEMORY_SIZE': settings.FILE_UPLOAD_MAX_MEMORY_SIZE,
    # Do not finish the url with a slash
    'MBGLRENDERER_URL': 'http://mbglrenderer',
    # We should automatically get the source of layers from a model
    'map': {
        "mapbox_access_token": None,
    },
    'STYLES': {
        'line': {
            'type': 'line',
            'paint': {
                'line-color': '#000',
                'line-width': 3
            }
        },
        'point': {
            'type': 'circle',
            'paint': {
                'circle-color': '#000',
                'circle-radius': 8
            }
        },
        'polygon': {
            'type': 'fill',
            'paint': {
                'fill-color': '#000'
            }
        },
    },
    'MAX_ZOOM': 15
}
_DEFAULT_TERRA_GEOCRUD.update(getattr(settings, 'TERRA_GEOCRUD', {}))
TERRA_GEOCRUD = deepcopy(_DEFAULT_TERRA_GEOCRUD)
